### PR TITLE
Experimenting with pre-processing output for plots

### PR DESCRIPTION
**Pull request by Karm**
* fixed formatter, it wasn't running on `verify` as it should have :( 
* adds `@Path("experiment/{image_name}")` for me to test and play with various plots on a client, makes my life easier in json.c
* fixed `gc_time_sec` so as total time the image build took and total time GC took are not confused

### CODE REVIEW

**Comment by Karm** [comment]
@jerboaa  

Demoed with this https://github.com/Karm/dev-null/blob/main/code-with-quarkus_3/run.sh#L4

I can get the ordered output of data series I wanted to experiment with:

```bash
$ curl -s -w '\n' -H "Content-Type: application/json" -H "token: $TOKEN" --get "$URL/experiment/q3-build-perf-karm-1.0.0-runner" | jq
```

```json
{
  "classes_jni": [
    61,
    63,
    61,
    63,
    63
  ],
  "classes_reachable": [
    21262,
    21605,
    21714,
    21605,
    21604
  ],
  "classes_reflection": [
    6294,
    808,
    805,
    808,
    808
  ],
  "classes_total": [
    23652,
    23548,
    23683,
    23547,
    23547
  ],
  "code_area_bytes": [
    40518000,
    37177456,
    36988688,
    37175744,
    37171408
  ],
  "fields_jni": [
    59,
    68,
    60,
    68,
    68
  ],
  "fields_reachable": [
    29313,
    29247,
    29295,
    29247,
    29247
  ],
  "fields_reflection": [
    431,
    408,
    424,
    408,
    408
  ],
  "fields_total": [
    46468,
    46236,
    46365,
    46237,
    46239
  ],
  "gc_total_ms": [
    6161,
    5354,
    5580,
    5605,
    7182
  ],
  "image_heap_bytes": [
    40337408,
    39272448,
    38301696,
    39272448,
    39268352
  ],
  "image_total_bytes": [
    81256872,
    76777344,
    75622600,
    76786496,
    76774160
  ],
  "methods_jni": [
    55,
    55,
    55,
    55,
    55
  ],
  "methods_reachable": [
    101167,
    100350,
    100860,
    100350,
    100348
  ],
  "methods_reflection": [
    4189,
    4083,
    4087,
    4083,
    4083
  ],
  "methods_total": [
    182491,
    183225,
    184640,
    183206,
    183214
  ],
  "peak_rss_bytes": [
    7484608512,
    7604346880,
    7713759232,
    7500648448,
    7545528320
  ],
  "resources_bytes": [
    428722,
    423166,
    427241,
    423166,
    423166
  ],
  "resources_count": [
    29,
    29,
    29,
    29,
    29
  ],
  "tag": [
    "quay.io/karmkarm/ubi-quarkus-mandrel-builder-image:23.0-java20",
    "quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.3.1-java17",
    "quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.3.1-java19",
    "quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.3.1.0-Final-java17",
    "quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.3.2.0-Final-java17"
  ],
  "total_build_time_ms": [
    111780,
    121143,
    122286,
    138814,
    139481
  ]
}
```

I know I mentioned somewhere that the number of resources included differed and here it looks like it doesn't. Apparently it depends on the test app.  
I had this in my notes.txt
```
Mandrel
  31.99MB (48.95%) for code area:    52,916 compilation units
  33.04MB (50.56%) for image heap:  387,418 objects and 160 resources
GraalVM
  31.99MB (49.72%) for code area:    52,916 compilation units
  32.04MB (49.80%) for image heap:  380,501 objects and 82 resources
```
regarding MicroProfile showcase app, so I will revisit it and confirm....

**Comment by Karm** [comment]
@jerboaa classes_reflection :scream: for `quay.io/karmkarm/ubi-quarkus-mandrel-builder-image:23.0-java20` seems like a bug in the collection.

## 22.3 JDK17
```
  21,605 (91.75%) of 23,548 classes reachable
  29,247 (63.26%) of 46,236 fields reachable
 100,350 (54.77%) of 183,225 methods reachable
     808 classes,   408 fields, and 4,083 methods registered for reflection
      63 classes,    68 fields, and    55 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
```

## 23.0 JDK20
```
  21,262 (89.90%) of 23,652 types reachable
  29,313 (63.08%) of 46,468 fields reachable
 101,167 (55.44%) of 182,491 methods reachable
   6,294 types,   431 fields, and 4,189 methods registered for reflection
      61 types,    59 fields, and    55 methods registered for JNI access
       4 native libraries: dl, pthread, rt, z
```

So both `classes registered for reflection` and `types registered for reflection` end up collected in [numClasses](https://github.com/Karm/collector/blob/main/src/main/java/com/redhat/quarkus/mandrel/collector/report/model/ClassStats.java#L46)

**Comment by jerboaa** [comment]
> @jerboaa classes_reflection scream for `quay.io/karmkarm/ubi-quarkus-mandrel-builder-image:23.0-java20` seems like a bug in the collection.
> ## 22.3 JDK17
> 
> ```
>   21,605 (91.75%) of 23,548 classes reachable
>   29,247 (63.26%) of 46,236 fields reachable
>  100,350 (54.77%) of 183,225 methods reachable
>      808 classes,   408 fields, and 4,083 methods registered for reflection
>       63 classes,    68 fields, and    55 methods registered for JNI access
>        4 native libraries: dl, pthread, rt, z
> ```
> 
> ## 23.0 JDK20
> 
> ```
>   21,262 (89.90%) of 23,652 types reachable
>   29,313 (63.08%) of 46,468 fields reachable
>  101,167 (55.44%) of 182,491 methods reachable
>    6,294 types,   431 fields, and 4,189 methods registered for reflection
>       61 types,    59 fields, and    55 methods registered for JNI access
>        4 native libraries: dl, pthread, rt, z
> ```
> 
> So both `classes registered for reflection` and `types registered for reflection` end up collected in [numClasses](https://github.com/Karm/collector/blob/main/src/main/java/com/redhat/quarkus/mandrel/collector/report/model/ClassStats.java#L46)

I'm not following, yes they are. The old one was called "classes reachable" the new one is "types reachable". See https://github.com/Karm/collector/commit/fbc5cee6edee9b501ec6044e4581270de4e00965

**Comment by jerboaa** [comment]
> What would be the apples to apples normalization to plot 22 and 23 side by side then? Is it really just a name with the same semantics?

For starters, use the same base JDK. You are comparing JDK 17 with JDK 20, no? I keep digging if the implementation changed as well.

File: src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java [hunk 1]
```diff
@@ -79,6 +83,63 @@ public ImageStats[] getByTag(@PathParam("tag") String tag) {
         return collection.getAllByTag(tag);
     }
 
+    @RolesAllowed("token_read")
+    @GET
+    @Path("experiment/{image_name}")
+    public Response getExperiment(@PathParam("image_name") String imageName) {
+        final ImageStats[] s = collection.getAllByImageName(imageName);
```
Comment (jerboaa) [inline src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java:90]: We could consider exposing this as `@Path("name/{image_name}"` as well. Seems useful. But, another PR :)

File: src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java [hunk 2]
```diff
@@ -79,6 +83,63 @@ public ImageStats[] getByTag(@PathParam("tag") String tag) {
...
+            nums.get("classes_total").add(i.getTotalStats().getNumClasses());
+            nums.get("classes_reflection").add(i.getReflectionStats().getNumClasses());
+            nums.get("classes_jni").add(i.getJniStats().getNumClasses());
+            nums.get("classes_reachable").add(i.getReachableStats().getNumClasses());
+            nums.get("fields_total").add(i.getTotalStats().getNumFields());
+            nums.get("fields_reflection").add(i.getReflectionStats().getNumFields());
+            nums.get("fields_jni").add(i.getJniStats().getNumFields());
+            nums.get("fields_reachable").add(i.getReachableStats().getNumFields());
+        }
+        final JsonObjectBuilder rb = Json.createObjectBuilder();
```
Comment (jerboaa) [inline src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java:136]: Eventually, we should model this with Java classes. This looks pretty ugly.

**Review by jerboaa — APPROVED** [review]
Seems fine.

File: src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java (hunk 2 above)
Comment (Karm) [reply src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java:136]: I agree. I merely want to "script", for a lack of a better word, prototype to play with I can comfortable consume and plot and tie triggers to. I would like to finally cross the gap between collecting the data and having a simple, meaningful insight into "something is wrong".

I will refactor it into a POJO that could also be e.g. cached etc., later...

Comment (Karm) [reply src/main/java/com/redhat/quarkus/mandrel/collector/report/endpoints/ImageStatsResource.java:136]: At some point I would like to generalize it so as you can "start an experiment", e.g. with a specific Quarkus (or other) app, get Jenkins collect a whole matrix of platforms and also flags, settings etc., plot the whole result and terminate the experiment with some conclusion.

**Pull request merged**